### PR TITLE
feat(presentation): proper schema

### PR DESCRIPTION
**Pull request by aldy505**
schema was by Midtrans, all rights reserved

there are some TODOs, others can contribute for it. it's easy enough, I think.

I intend to make this repository as close to Midtrans and how they handle the logic (and the webhook/notification), so when we got a future project that requires integration with Midtrans, we already did save some time developing it.

### CODE REVIEW

**Review by YogiPristiawan — APPROVED** [review]

File: presentation/charge_transaction.go [hunk 1]
```diff
@@ -31,38 +31,68 @@ func (p *Presenter) ChargeTransaction(w http.ResponseWriter, r *http.Request) {
...
 	}
 
+	paymentType, err := parseValidPaymentMethod(requestBody)
+	if err != nil {
+		responseBody, e := json.Marshal(schema.Error{
+			StatusCode:    http.StatusBadRequest,
+			StatusMessage: err.Error(),
+		})
+		if e != nil {
+			w.WriteHeader(http.StatusInternalServerError)
```
Comment (WahidinAji) [inline presentation/charge_transaction.go:41]: karena abis kejdian marshar error ga nampilin body. gatau kenapa ini keknya perlu deh, walaupun cuma simple 

w.Write([]byte("internal server error"))


kasihan FE kalo cuma baca header. soale kemarin kejadian di saya FE nya komplen wkwkkwkwkwk ;D

File: presentation/charge_transaction.go [hunk 2]
```diff
@@ -31,38 +31,68 @@ func (p *Presenter) ChargeTransaction(w http.ResponseWriter, r *http.Request) {
...
-				Address:     requestBody.Customer.BillingAddress.Address,
-				PostalCode:  requestBody.Customer.BillingAddress.PostalCode,
-				CountryCode: requestBody.Customer.BillingAddress.CountryCode,
+				FirstName:   requestBody.CustomerDetails.BillingAddress.FirstName,
+				LastName:    requestBody.CustomerDetails.BillingAddress.LastName,
+				Email:       requestBody.CustomerDetails.BillingAddress.Email,
+				Phone:       requestBody.CustomerDetails.BillingAddress.Phone,
+				Address:     requestBody.CustomerDetails.BillingAddress.Address,
+				PostalCode:  requestBody.CustomerDetails.BillingAddress.PostalCode,
+				CountryCode: requestBody.CustomerDetails.BillingAddress.CountryCode,
```
Comment (WahidinAji) [inline presentation/charge_transaction.go:76]:
```suggestion
				FirstName:   requestBody.CustomerDetail.BillingAddress.FirstName,
				LastName:    requestBody.CustomerDetail.BillingAddress.LastName,
				Email:       requestBody.CustomerDetail.BillingAddress.Email,
				Phone:       requestBody.CustomerDetail.BillingAddress.Phone,
				Address:     requestBody.CustomerDetail.BillingAddress.Address,
				PostalCode:  requestBody.CustomerDetail.BillingAddress.PostalCode,
				CountryCode: requestBody.CustomerDetail.BillingAddress.CountryCode,
```

File: presentation/charge_transaction.go [hunk 3]
```diff
@@ -31,38 +31,68 @@ func (p *Presenter) ChargeTransaction(w http.ResponseWriter, r *http.Request) {
...
+		TransactionCurrency: currencyMap[requestBody.TransactionDetails.Currency],
 		Customer: business.CustomerInformation{
-			FirstName:   requestBody.Customer.FirstName,
-			LastName:    requestBody.Customer.LastName,
-			Email:       requestBody.Customer.Email,
-			PhoneNumber: requestBody.Customer.PhoneNumber,
+			FirstName:   requestBody.CustomerDetails.FirstName,
+			LastName:    requestBody.CustomerDetails.LastName,
+			Email:       requestBody.CustomerDetails.Email,
+			PhoneNumber: requestBody.CustomerDetails.PhoneNumber,
```
Comment (WahidinAji) [inline presentation/charge_transaction.go:68]: ini single kan ya? harusnya singular ga? misal `requestBody.CustomerDetail.FirstName` 

```suggestion
			FirstName:   requestBody.CustomerDetail.FirstName,
			LastName:    requestBody.CustomerDetail.LastName,
			Email:       requestBody.CustomerDetail.Email,
			PhoneNumber: requestBody.CustomerDetail.PhoneNumber,
```


soalnya kuliat disini single customer aja kan? https://github.com/TokoBapak/mock-payment-provider/pull/17/files#

File: presentation/charge_transaction.go [hunk 4]
```diff
@@ -157,21 +187,204 @@ func (p *Presenter) ChargeTransaction(w http.ResponseWriter, r *http.Request) {
...
+			PaymentType:            "",
+			TransactionTime:        "",
+			TransactionStatus:      "",
+			Actions:                nil,
+			ChannelResponseCode:    "",
+			ChannelResponseMessage: "",
+			Currency:               "",
+		})
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
```
Comment (WahidinAji) [inline presentation/charge_transaction.go:207]: ini juga, `w.Write([]byte("internal server error")` misal

File: presentation/charge_transaction.go [hunk 5]
```diff
@@ -157,21 +187,204 @@ func (p *Presenter) ChargeTransaction(w http.ResponseWriter, r *http.Request) {
...
+			GrossAmount:            "",
+			Currency:               "",
+			PaymentType:            "",
+			TransactionTime:        "",
+			TransactionStatus:      "",
+			FraudStatus:            "",
+			Actions:                nil,
+		})
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
```
Comment (WahidinAji) [inline presentation/charge_transaction.go:233]: 3. w.Write

File: presentation/charge_transaction.go [hunk 6]
```diff
@@ -157,21 +187,204 @@ func (p *Presenter) ChargeTransaction(w http.ResponseWriter, r *http.Request) {
...
+			Currency:          "",
+			PaymentType:       "",
+			TransactionTime:   "",
+			TransactionStatus: "",
+			FraudStatus:       "",
+			Acquirer:          "",
+			Actions:           nil,
+		})
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
```
Comment (WahidinAji) [inline presentation/charge_transaction.go:258]: semua lahh pokoknya kwkwkwkwkwwk

**Review by WahidinAji — CHANGES_REQUESTED** [review]

File: presentation/charge_transaction.go (hunk 1 above)
Comment (aldy505) [reply presentation/charge_transaction.go:41]: kan yang consume ini sesama backend...

File: presentation/charge_transaction.go (hunk 2 above)
Comment (aldy505) [reply presentation/charge_transaction.go:76]: nggak mau. ini kan targetnya mau 1:1 replica sama API nya midtrans. di midtrans dipakenya `CustomerDetails` bukan `CustomerDetail`.

kalo dibikin sesuai kaidah yang wajar, nggak jadi 1:1 replica.

File: presentation/charge_transaction.go (hunk 3 above)
Comment (aldy505) [reply presentation/charge_transaction.go:68]: nggak mau. ini kan targetnya mau 1:1 replica sama API nya midtrans. di midtrans dipakenya CustomerDetails bukan CustomerDetail.

kalo dibikin sesuai kaidah yang wajar, nggak jadi 1:1 replica.

File: presentation/charge_transaction.go (hunk 4 above)
Comment (aldy505) [reply presentation/charge_transaction.go:207]: nggak mau. biar sesuai ngikutin midtrans aja.

File: presentation/charge_transaction.go (hunk 5 above)
Comment (aldy505) [reply presentation/charge_transaction.go:233]: nggak mau. biar sesuai ngikutin midtrans aja.

File: presentation/charge_transaction.go (hunk 6 above)
Comment (aldy505) [reply presentation/charge_transaction.go:258]: nggak mau. biar sesuai ngikutin midtrans aja.

**Pull request merged**